File: node/src/proofOfCloud.ts
import { readFileSync, writeFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { join, dirname } from "node:path";
import { AttestationResult, makeResult, orderChecks } from "./types.js";
import { detectCpuQuoteType } from "./cpu.js";

const PEERS_GITHUB_RAW =
  "https://raw.githubusercontent.com/proofofcloud/trust-server/main/public_info/peers_list.txt";

function peersFilePath(): string {
  const __filename = fileURLToPath(import.meta.url);
  const __dirname = dirname(__filename);
  return join(__dirname, "..", "data", "trust_server_peers.txt");
}

// ---------------------------------------------------------------------------
// Quote -> canonical lowercase hex
// ---------------------------------------------------------------------------

/**
 * Encode a CPU quote (as served by the VM's /cpu endpoint) to a canonical
 * lowercase hex string suitable for the trust-server `/check_quote` API.
 *
 * Reuses the existing quote-type detector in cpu.ts (hex => TDX,
 * base64 => SEV-SNP) but then strictly validates the WHOLE string:
 *
 *   - TDX: the trimmed string must match /^[0-9a-fA-F]+$/ with even length.
 *     (We do NOT round-trip through Buffer.from(text, "hex"), which silently
 *     truncates at the first invalid character.)
 *   - SEV-SNP: strict base64 decode (no silent truncation), then re-encode the
 *     decoded bytes as lowercase hex.
 *
 * Throws on unrecognized / invalid input so the orchestrator can early-return
 * without making any network call.
 */
export function toHexQuote(cpuData: string): string {
  const text = cpuData.trim();
  const quoteType = detectCpuQuoteType(text);

  if (quoteType === "TDX") {
    if (text.length % 2 !== 0 || !/^[0-9a-fA-F]+$/.test(text)) {
      throw new Error("invalid TDX hex quote");
    }
    return text.toLowerCase();
  }

  if (quoteType === "SEV-SNP") {
    const buf = Buffer.from(text, "base64");
    // Strict base64 check: re-encoding the decoded bytes must reproduce the
    // input (modulo padding/whitespace), otherwise base64 silently dropped
    // invalid characters.
    const normalized = text.replace(/\s+/g, "");
    const reencoded = buf.toString("base64");
    const stripPad = (s: string) => s.replace(/=+$/, "");
    if (buf.length === 0 || stripPad(reencoded) !== stripPad(normalized)) {
      throw new Error("invalid SEV-SNP base64 quote");
    }
    return buf.toString("hex").toLowerCase();
  }

  throw new Error("unrecognized quote format");
}

// ---------------------------------------------------------------------------
// Peer list loading + refresh
// ---------------------------------------------------------------------------

/** Parse newline-delimited peer text into a list of https origins. */
function parsePeers(content: string): string[] {
  const peers: string[] = [];
  for (const rawLine of content.split("\n")) {
    const line = rawLine.trim();
    if (!line || line.startsWith("#")) continue;
    try {
      const u = new URL(line);
      if (u.protocol !== "https:") continue;
      peers.push(u.origin);
    } catch {
      // drop invalid line individually
    }
  }
  return peers;
}

/** Read and parse the bundled peers file (offline fallback). */
export function loadBundledPeers(): string[] {
  try {
    return parsePeers(readFileSync(peersFilePath(), "utf8"));
  } catch {
    return [];
  }
}

/**
 * Fetch the raw peers list from GitHub with a 5s timeout. Returns the parsed
 * peers if >=1 valid, else null. Best-effort persists the fetched raw text to
 * the bundled file (write errors are ignored). Never throws.
 */
export async function refreshPeersFromGitHub(): Promise<string[] | null> {
  try {
    const resp = await fetch(PEERS_GITHUB_RAW, {
      signal: AbortSignal.timeout(5000),
    });
    if (!resp.ok) return null;
    const text = await resp.text();
    const peers = parsePeers(text);
    if (peers.length < 1) return null;
    // best-effort persist for future offline starts
    try {
      writeFileSync(peersFilePath(), text, "utf8");
    } catch {
      // read-only install (Docker, global npm) — ignore
    }
    return peers;
  } catch {
    return null;
  }
}

let _peersCache: string[] | null = null;

/**
 * Resolve the peer list for this process. On first call, attempt one GitHub
 * refresh; use it if it yields >=1 peer, else fall back to the bundled file.
 * Memoized at module scope for the rest of the process.
 */
export async function resolvePeers(): Promise<string[]> {
  if (_peersCache) return _peersCache;
  const refreshed = await refreshPeersFromGitHub();
  _peersCache = refreshed ?? loadBundledPeers();
  return _peersCache;
}

/** Reset the module-level peers memoization (for tests). */
export function resetPeersCacheForTests(): void {
  _peersCache = null;
}

// ---------------------------------------------------------------------------
// Single-peer query
// ---------------------------------------------------------------------------

interface PeerAnswer {
  whitelisted: boolean;
  machine_id: string;
  revoked: boolean;
  revoked_at: string | null;
}

/**
 * Query one trust-server peer. Returns a parsed usable answer or a failure
 * reason string. A usable answer is HTTP 200 + JSON where `whitelisted` is a
 * boolean and `machine_id` is a non-empty string.
 */
async function queryPeer(
  peerUrl: string,
  hexQuote: string,
): Promise<PeerAnswer | string> {
  let resp: Response;
  try {
    resp = await fetch(`${peerUrl}/check_quote`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ quote: hexQuote }),
      signal: AbortSignal.timeout(10000),
    });
  } catch (e: any) {
    return `${peerUrl}: request failed (${e?.message ?? e})`;
  }

  if (!resp.ok) {
    return `${peerUrl}: HTTP ${resp.status}`;
  }

  let body: any;
  try {
    body = await resp.json();
  } catch {
    return `${peerUrl}: response was not valid JSON`;
  }

  if (
    typeof body?.whitelisted !== "boolean" ||
    typeof body?.machine_id !== "string" ||
    body.machine_id.length === 0
  ) {
    return `${peerUrl}: unexpected response shape`;
  }

  let revoked = false;
  if (body.revoked !== undefined) {
    if (typeof body.revoked !== "boolean") {
      return `${peerUrl}: unexpected response shape (revoked)`;
    }
    revoked = body.revoked;
  }

  const revoked_at =
    revoked && typeof body.revoked_at === "string" ? body.revoked_at : null;

  return {
    whitelisted: body.whitelisted,
    machine_id: body.machine_id,
    revoked,
    revoked_at,
  };
}

// ---------------------------------------------------------------------------
// Orchestrator
// ---------------------------------------------------------------------------

function buildReport(over: Partial<Record<string, any>>): Record<string, any> {
  return {
    proof_of_cloud: {
      whitelisted: false,
      machine_id: null,
      revoked: false,
      revoked_at: null,
      trust_server: null,
      peers_tried: [],
      ...over,
    },
  };
}

/**
 * Verify a CPU quote against the community-vetted trust-server peers.
 *
 * The quote is encoded to canonical lowercase hex and POSTed to each peer's
 * `/check_quote` endpoint in list order until one returns a usable answer
 * (failover; first usable answer wins). The verdict passes iff the accepted
 * answer is `whitelisted` and not `revoked`.
 *
 * `report.proof_of_cloud` is always populated on every return path.
 */
export async function checkProofOfCloud(
  cpuData: string,
): Promise<AttestationResult> {
  const checks: Record<string, boolean> = {};
  const errors: string[] = [];

  // 1. Encode the quote to hex (no network call on failure).
  let hexQuote: string;
  try {
    hexQuote = toHexQuote(cpuData);
  } catch {
    errors.push(
      "Could not encode quote for trust-server (unrecognized quote format)",
    );
    checks.proof_of_cloud_verified = false;
    return makeResult("PROOF-OF-CLOUD", {
      checks: orderChecks(checks),
      report: buildReport({ peers_tried: [] }),
      errors,
    });
  }

  // 2. Resolve peers.
  const peers = await resolvePeers();
  if (peers.length === 0) {
    errors.push("No trust-server peers available");
    checks.proof_of_cloud_verified = false;
    return makeResult("PROOF-OF-CLOUD", {
      checks: orderChecks(checks),
      report: buildReport({ peers_tried: [] }),
      errors,
    });
  }

  // 3. Failover: try peers in order until first usable answer.
  const peersTried: string[] = [];
  const reasons: string[] = [];
  let answer: PeerAnswer | null = null;
  let trustServer: string | null = null;

  for (const peer of peers) {
    peersTried.push(peer);
    const res = await queryPeer(peer, hexQuote);
    if (typeof res === "string") {
      reasons.push(res);
      continue;
    }
    answer = res;
    trustServer = peer;
    break;
  }

  // 4. No usable answer from any peer.
  if (!answer || !trustServer) {
    for (const r of reasons) errors.push(r);
    checks.proof_of_cloud_verified = false;
    return makeResult("PROOF-OF-CLOUD", {
      checks: orderChecks(checks),
      report: buildReport({ peers_tried: peersTried }),
      errors,
    });
  }

  // 5. Build verdict from the accepted answer.
  if (answer.revoked) {
    errors.push(
      `Machine ${answer.machine_id} was revoked on ${answer.revoked_at}`,
    );
    checks.proof_of_cloud_verified = false;
    return makeResult("PROOF-OF-CLOUD", {
      checks: orderChecks(checks),
      report: buildReport({
        whitelisted: false,
        machine_id: answer.machine_id,
        revoked: true,
        revoked_at: answer.revoked_at,
        trust_server: trustServer,
        peers_tried: peersTried,
      }),
      errors,
    });
  }

  if (!answer.whitelisted) {
    errors.push(
      `Machine ${answer.machine_id} is not whitelisted by trust-server peer ${trustServer}`,
    );
    checks.proof_of_cloud_verified = false;
    return makeResult("PROOF-OF-CLOUD", {
      checks: orderChecks(checks),
      report: buildReport({
        whitelisted: false,
        machine_id: answer.machine_id,
        revoked: false,
        revoked_at: null,
        trust_server: trustServer,
        peers_tried: peersTried,
      }),
      errors,
    });
  }

  // Pass.
  checks.proof_of_cloud_verified = true;
  return makeResult("PROOF-OF-CLOUD", {
    valid: true,
    checks: orderChecks(checks),
    report: buildReport({
      whitelisted: true,
      machine_id: answer.machine_id,
      revoked: false,
      revoked_at: null,
      trust_server: trustServer,
      peers_tried: peersTried,
    }),
    errors,
  });
}

File: python/src/secretvm/verify/proof_of_cloud.py
"""Proof-of-Cloud verification against the trust-server peer network.

The verdict is produced by failing over across the community-vetted
trust-server peers (https://github.com/proofofcloud/trust-server). Each peer
exposes ``POST /check_quote`` taking a hex-encoded quote and returning
``{whitelisted, machine_id, revoked?, revoked_at?}``. Peers are tried in list
order; the first usable answer wins. The peer list ships bundled and is
best-effort refreshed from GitHub once per process.
"""

import asyncio
import base64
import re
import urllib.parse
import urllib.request
from pathlib import Path
from typing import Optional

import requests

from .cpu import _detect_cpu_quote_type
from .types import AttestationResult, order_checks

# Raw GitHub URL for the published peers list.
_PEERS_GITHUB_RAW = (
    "https://raw.githubusercontent.com/proofofcloud/trust-server/main/"
    "public_info/peers_list.txt"
)

_HEX_RE = re.compile(r"^[0-9a-fA-F]+$")

# Module-level memoization of the resolved peer list (per process).
_peers_cache: Optional[list] = None


# ---------------------------------------------------------------------------
# Quote -> hex encoding
# ---------------------------------------------------------------------------

def to_hex_quote(cpu_data: str) -> str:
    """Return the canonical lowercase-hex encoding of a CPU quote.

    Reuses the existing TDX/SEV-SNP detection, then strictly validates the
    *whole* trimmed string before returning (the detectors only sniff header
    bytes, and ``bytes.fromhex`` / base64 decode can silently accept partially
    malformed input).

    - TDX (hex): the entire trimmed string must match ``^[0-9a-fA-F]+$`` with
      even length; the result is lowercased.
    - SEV-SNP (base64): strict base64 decode of the whole string, re-encoded as
      lowercase hex.

    Raises:
        ValueError: if the input matches neither a valid hex nor a valid base64
            quote (no network call should be made by the caller in that case).
    """
    text = cpu_data.strip()
    quote_type = _detect_cpu_quote_type(text)

    if quote_type == "TDX":
        if len(text) % 2 != 0 or not _HEX_RE.match(text):
            raise ValueError(
                "Could not encode quote for trust-server (unrecognized quote format)"
            )
        return text.lower()

    if quote_type == "SEV-SNP":
        # Strip internal whitespace first (a VM may serve line-wrapped base64),
        # matching the Node SDK and the main SEV path in amd.py, then strict-decode.
        compact = re.sub(r"\s+", "", text)
        try:
            raw = base64.b64decode(compact, validate=True)
        except Exception:
            raise ValueError(
                "Could not encode quote for trust-server (unrecognized quote format)"
            )
        return raw.hex()

    raise ValueError(
        "Could not encode quote for trust-server (unrecognized quote format)"
    )


# ---------------------------------------------------------------------------
# Peer list loader + refresher
# ---------------------------------------------------------------------------

def _parse_peers(text: str) -> list:
    """Parse peer-list text into a list of https origins.

    Split on newlines; trim each line; skip blanks and ``#`` comments. Parse
    each remaining line as a URL, keep only ``https`` scheme, and reduce to its
    origin (``scheme://netloc``). Invalid or non-https lines are dropped
    individually.
    """
    peers: list = []
    for line in text.splitlines():
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        try:
            parsed = urllib.parse.urlparse(line)
        except Exception:
            continue
        if parsed.scheme != "https" or not parsed.netloc:
            continue
        origin = f"{parsed.scheme}://{parsed.netloc}"
        peers.append(origin)
    return peers


def load_bundled_peers() -> list:
    """Read and parse the bundled peers file (offline fallback)."""
    path = Path(__file__).parent / "data" / "trust_server_peers.txt"
    try:
        text = path.read_text(encoding="utf-8")
    except Exception:
        return []
    return _parse_peers(text)


def refresh_peers_from_github() -> Optional[list]:
    """Fetch the published peers list from GitHub (5s timeout, best-effort).

    Returns the parsed peer list if it yields >=1 valid peer, else ``None``.
    Best-effort persists the fetched raw text to the bundled file (write errors
    ignored). Never raises.
    """
    try:
        with urllib.request.urlopen(_PEERS_GITHUB_RAW, timeout=5) as resp:
            text = resp.read().decode("utf-8")
    except Exception:
        return None

    peers = _parse_peers(text)
    if not peers:
        return None

    # Best-effort persist so a future offline process starts from the newer list.
    try:
        path = Path(__file__).parent / "data" / "trust_server_peers.txt"
        path.write_text(text, encoding="utf-8")
    except Exception:
        pass

    return peers


def _resolve_peers() -> list:
    """Module-memoized peer resolution.

    On first call, attempt one GitHub refresh; if it yields >=1 valid peer use
    it, else fall back to the bundled file. The result is cached for the rest of
    the process.
    """
    global _peers_cache
    if _peers_cache is not None:
        return _peers_cache

    refreshed = refresh_peers_from_github()
    if refreshed:
        _peers_cache = refreshed
    else:
        _peers_cache = load_bundled_peers()
    return _peers_cache


def _reset_peers_cache() -> None:
    """Clear the memoized peer list (for tests)."""
    global _peers_cache
    _peers_cache = None


# ---------------------------------------------------------------------------
# Single-peer query
# ---------------------------------------------------------------------------

def _query_peer(origin: str, hex_quote: str):
    """Query a single peer's ``/check_quote`` endpoint.

    Returns a dict ``{whitelisted, machine_id, revoked, revoked_at}`` on a
    usable answer, or a failure-reason string otherwise.
    """
    try:
        resp = requests.post(
            f"{origin}/check_quote",
            json={"quote": hex_quote},
            timeout=10,
        )
    except Exception as e:
        return f"{origin}: request failed: {e}"

    if resp.status_code != 200:
        return f"{origin}: HTTP {resp.status_code}"

    try:
        body = resp.json()
    except Exception:
        return f"{origin}: response was not valid JSON"

    if not isinstance(body, dict):
        return f"{origin}: response was not a JSON object"

    whitelisted = body.get("whitelisted")
    machine_id = body.get("machine_id")
    if not isinstance(whitelisted, bool):
        return f"{origin}: 'whitelisted' missing or not a boolean"
    if not isinstance(machine_id, str) or not machine_id:
        return f"{origin}: 'machine_id' missing or not a non-empty string"

    revoked_raw = body.get("revoked")
    if revoked_raw is not None and not isinstance(revoked_raw, bool):
        return f"{origin}: 'revoked' present but not a boolean"
    revoked = bool(revoked_raw)

    revoked_at = None
    if revoked:
        ra = body.get("revoked_at")
        revoked_at = ra if isinstance(ra, str) else None

    return {
        "whitelisted": whitelisted,
        "machine_id": machine_id,
        "revoked": revoked,
        "revoked_at": revoked_at,
    }


# ---------------------------------------------------------------------------
# Orchestrator
# ---------------------------------------------------------------------------

def check_proof_of_cloud(cpu_data: str) -> AttestationResult:
    """Verify a CPU quote against the trust-server peer network.

    Encodes the quote to hex, resolves the peer list, then queries peers in
    order until the first usable answer. Passes iff that answer is
    ``whitelisted`` and not ``revoked``.

    Args:
        cpu_data: Raw CPU quote text as served by the VM's /cpu endpoint
            (hex for TDX, base64 for SEV-SNP).

    Returns:
        AttestationResult with attestation_type="PROOF-OF-CLOUD". The single
        check ``proof_of_cloud_verified`` reflects the verdict. ``report
        ["proof_of_cloud"]`` is always populated.
    """
    checks: dict = {}
    errors: list = []

    def _result(valid: bool, *, whitelisted=False, machine_id=None, revoked=False,
                revoked_at=None, trust_server=None, peers_tried=None) -> AttestationResult:
        checks["proof_of_cloud_verified"] = valid
        report = {
            "proof_of_cloud": {
                "whitelisted": whitelisted,
                "machine_id": machine_id,
                "revoked": revoked,
                "revoked_at": revoked_at,
                "trust_server": trust_server,
                "peers_tried": peers_tried if peers_tried is not None else [],
            }
        }
        return AttestationResult(
            valid=valid, attestation_type="PROOF-OF-CLOUD",
            checks=order_checks(checks), report=report, errors=errors,
        )

    # 1. Encode the quote to canonical hex. Encode failure => no network call.
    try:
        hex_quote = to_hex_quote(cpu_data)
    except ValueError as e:
        errors.append(str(e))
        return _result(False, peers_tried=[])

    # 2. Resolve the peer list.
    peers = _resolve_peers()
    if not peers:
        errors.append("No trust-server peers available")
        return _result(False, peers_tried=[])

    # 3. Failover: try peers in order, first usable answer wins. Per-peer
    # transport/parse failures are collected separately and only surfaced if no
    # peer returns a usable answer (a definitive verdict from a later peer must
    # not carry earlier peers' failure noise).
    peers_tried: list = []
    reasons: list = []
    for origin in peers:
        peers_tried.append(origin)
        answer = _query_peer(origin, hex_quote)
        if isinstance(answer, str):
            reasons.append(answer)
            continue

        machine_id = answer["machine_id"]
        whitelisted = answer["whitelisted"]
        revoked = answer["revoked"]
        revoked_at = answer["revoked_at"]

        if revoked:
            when = revoked_at if revoked_at else "an unknown date"
            errors.append(f"Machine {machine_id} was revoked on {when}")
            return _result(
                False, whitelisted=False, machine_id=machine_id, revoked=True,
                revoked_at=revoked_at, trust_server=origin, peers_tried=peers_tried,
            )

        if not whitelisted:
            errors.append(
                f"Machine {machine_id} is not whitelisted by trust-server peer {origin}"
            )
            return _result(
                False, whitelisted=False, machine_id=machine_id, revoked=False,
                revoked_at=None, trust_server=origin, peers_tried=peers_tried,
            )

        # Pass.
        return _result(
            True, whitelisted=True, machine_id=machine_id, revoked=False,
            revoked_at=None, trust_server=origin, peers_tried=peers_tried,
        )

    # 4. No peer returned a usable answer — surface every peer's failure reason.
    errors.extend(reasons)
    return _result(False, peers_tried=peers_tried)


async def check_proof_of_cloud_async(cpu_data: str) -> AttestationResult:
    """Async variant of check_proof_of_cloud (offloads the blocking path)."""
    return await asyncio.to_thread(check_proof_of_cloud, cpu_data)
